Upgrade to bindings 0.6.0 and SDK 0.12.20

stefanobaghino-da opened the pull request:
- use the latest and greatest bindings
  - showcase the usage of helpers to express DAML values more concisely†
- use the latest and greatest SDK
  - replace usages of the `da` SDK assistant with `daml`
  - switch the default port to 6865 (instead of 7600)
- improve the package manifest and make it more complete
- fix layout errors in the `README.md`
- remove references to older APIs
- fix error with transactions and explain the rationale

† unfortunately in this example this doesn't yield much savings, but it's still worth to show it, I believe

stefanobaghino-da committed:
Upgrade to bindings 0.6.0 and SDK 0.12.20

gerolf-da commented on README.md:
```diff
@@ -27,6 +29,14 @@ The focus is not on the complexity of the model, but rather on how to use the bi
+
+The DAML code in this repository makes use of the SDK 0.12.20 and the `daml` SDK assistant.
+
+You are therefore strongly advised to have at least the SDK 0.12.20 installed.
```
I think this should be more than a strong advice, as it seems that SDK 0.12.20 is more like a strict requirement?

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/ex-tutorial-nodejs/pull/7/commits/0ebe6ca939825c0e9622d2ee6311d97dc371a9ad

gerolf-da commented on README.md:
```diff
@@ -35,21 +45,30 @@ To set it up, clone the repo, making sure to checkout the tag corresponding to t
 Use the sandbox to run and test your application.
 
-Now that you created the project and you're in its directory, start the sandbox by running:
+Now that you created the project and you're in its directory:
```
This phrasing seems odd, since the previous step already needed to happen within the project directory.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/ex-tutorial-nodejs/pull/7/commits/482a5ab5f421d587c5258115ef755a98863df47b

gerolf-da commented on README.md:
```diff
@@ -77,24 +96,28 @@ The code for the script you just ran is `index.js`.
 
     const uuidv4 = require('uuid/v4');
 
 The first line loads the bindings and allows you to refer to them through the `daml` object.
```
The object is now called `ledger`.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/ex-tutorial-nodejs/pull/7/commits/f20379f70347c217327b508d449245e34a8b9801

gerolf-da commented on README.md:
```diff
@@ -77,24 +96,28 @@ The code for the script you just ran is `index.js`.
 The first line loads the bindings and allows you to refer to them through the `daml` object.
 
-The second one introduces a dependency that is going to be later used to generate unique identifiers; no need to worry about it now.
+The second one creates a shorthand for the `daml` object in the `daml-ledger` library, that can be used to express DAML values in your code concisely.
```
I don't think it's considered bad style to repeat the word `line` instead of `one` here and in the next line.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/ex-tutorial-nodejs/pull/7/commits/b55298166fd9b835812c647937ea52a4fc56a264

gerolf-da commented on README.md:
```diff
@@ -77,24 +96,28 @@ The code for the script you just ran is `index.js`.
-The second one introduces a dependency that is going to be later used to generate unique identifiers; no need to worry about it now.
+The second one creates a shorthand for the `daml` object in the `daml-ledger` library, that can be used to express DAML values in your code concisely.
+
+The third one introduces a dependency that is going to be later used to generate unique identifiers; no need to worry about it now.
```
`... that is later going to be used ...`
or
`... that is going to be used later ...`
Since I'm not a native speaker, which one is better, @bethaitman?

gerolf-da commented on README.md:
```diff
@@ -322,7 +345,7 @@ This method takes the following request:
     filtersByParty[sender] = { inclusive: { templateIds: [PING, PONG] } };
     const request = {
-        begin: { boundary: da.LedgerOffset.Boundary.END },
+        begin: { boundary: ledger.LedgerOffsetBoundaryValue.BEGIN },
```
Is the switch from `END` to `BEGIN` on purpose?

stefanobaghino-da replied:
Yeah, I'll clarify as this is why I had to come up with the complex explanation afterwards, thanks.

stefanobaghino-da replied:
Hopefully better explained in https://github.com/digital-asset/ex-tutorial-nodejs/pull/7/commits/31c20f8b85cf929172f994519a377111f011865f

stefanobaghino-da replied:
Improved on with https://github.com/digital-asset/ex-tutorial-nodejs/pull/7/commits/7cbd34f6f64471bf6912614e32c97740d4aff4ac

gerolf-da commented on README.md:
```diff
@@ -509,9 +532,11 @@ Wrap this code into a new function `react` that takes a `workflowId` and an `eve
 
 Finally, pass the `react` function as a parameter to the only call of `listenForTransactions`.
 
+The app is almost ready to start, there's only one tweak to make: before writing the code to react to transactions, in order for us to observe the transactions while running a single application, the code you wrote subscribed to the ledger from it's beginning (`LedgerOffsetBoundaryValue.BEGIN`). The next time you'll run the application you'll instead _tail_ the ledger subscribing to the end of if ...
```
I don't quite understand what the first sentence after the `:` is trying to say.

stefanobaghino-da replied:
Hopefully better explained in https://github.com/digital-asset/ex-tutorial-nodejs/commit/31c20f8b85cf929172f994519a377111f011865f

stefanobaghino-da replied:
Improved on with https://github.com/digital-asset/ex-tutorial-nodejs/pull/7/commits/7cbd34f6f64471bf6912614e32c97740d4aff4ac

gerolf-da commented on README.md:
```diff
@@ -845,10 +872,13 @@ Note that the transaction filter was factored out as it can be shared. The final
     });
 
-Before running this you should start with a clean ledger to avoid being confused by the unprocessed contracts from previous examples.
+Before running this you should start with a clean ledger to avoid being confused by the unprocessed contracts from previous examples:
```
Maybe it's okay to make a more assertive statement like:
```
Before running the application, start with a clean ledger to avoid receiving unprocessed contracts from previous examples:
```

stefanobaghino-da replied:
Very good catch, thanks!

Addressed by https://github.com/digital-asset/ex-tutorial-nodejs/pull/7/commits/579220195e8ad201ccfe026a812abe5fd3d908d1

stefanobaghino-da replied:
Improved on with https://github.com/digital-asset/ex-tutorial-nodejs/pull/7/commits/7cbd34f6f64471bf6912614e32c97740d4aff4ac

gerolf-da reviewed (commented):
I like the `ledger.daml` object a lot. It removes quite a bit of noise 👍

stefanobaghino-da committed:
Address https://github.com/digital-asset/ex-tutorial-nodejs/pull/7#discussion_r287393607

stefanobaghino-da committed:
Address https://github.com/digital-asset/ex-tutorial-nodejs/pull/7#discussion_r287393972

stefanobaghino-da committed:
Address https://github.com/digital-asset/ex-tutorial-nodejs/pull/7#discussion_r287394589

stefanobaghino-da committed:
Address https://github.com/digital-asset/ex-tutorial-nodejs/pull/7#discussion_r287394845

stefanobaghino-da committed:
Address https://github.com/digital-asset/ex-tutorial-nodejs/pull/7#discussion_r287397995

stefanobaghino-da committed:
Address two review comments

- https://github.com/digital-asset/ex-tutorial-nodejs/pull/7#discussion_r287396220
- https://github.com/digital-asset/ex-tutorial-nodejs/pull/7#discussion_r287397166

stefanobaghino-da committed:
Improve on https://github.com/digital-asset/ex-tutorial-nodejs/commit/31c20f8b85cf929172f994519a377111f011865f

bitonic reviewed (changes requested):
* Why the repeated "back to the table of contents"?
* The tag `v0.6.0` is currently not in this repo, so I cannot test this tutorial. Would it be possible to first create this tag so that I can try out the tutorial first?

stefanobaghino-da commented:
@bitonic
> Why the repeated "back to the table of contents"?
So that you can easily navigate between chapters if you need to.
> The tag v0.6.0 is currently not in this repo, so I cannot test this tutorial. Would it be possible to first create this tag so that I can try out the tutorial first?
This would not make sense: I would then need to delete and re-make a tag for every review that make me introduce a change, plus another after GitHub introduces a new commit with the squash-and-merge process (which is what I would ultimately do). This would simply defeat the purpose of following all the tutorial, including this tag, since this is provided as a convenience for readers. You should simply assume that the commit introducing these changes will be tagged as `v0.6.0`, so please just test the tutorial using `HEAD`.

bitonic commented:
@stefanobaghino-da I'm confused -- can't you create the tag now and then replace it when you merge?

Note that I'm asking since generally speaking when reviewing such tutorial / guides the way I review them is by following them and making sure they work. Being able to execute each command exactly as given greatly aids this process.

bitonic commented on README.md:
```diff
@@ -189,7 +210,7 @@ Read those from the command line by editing the part where the arguments are rea
+    port = port || 6865;
     if (!sender || !receiver) {
         console.log('Missing sender and/or receiver arguments, exiting.');
         process.exit(-1);
```
Please add the full invocation with sender / receiving below as you do further down, e.g. `run start Alice Bob`

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/ex-tutorial-nodejs/pull/7/commits/05ee6db892c9d4f5979aa06a323c3810a22ff091

bitonic commented on README.md:
```diff
@@ -404,7 +435,7 @@ When you are done, your code should look like the following:
             const request = {
-                begin: { offsetType: 'boundary', boundary: daml.LedgerOffsetBoundaryValue.END },
+                begin: { offsetType: 'boundary', boundary: ledger.LedgerOffsetBoundaryValue.BEGIN },
                 filter: { filtersByParty: filtersByParty }
```
Below, you say

```
    Alice starts reading transactions.
    Created Ping contract from Alice to Bob.
    Transaction read: 1
```

but I actually see

```
Alice starts reading transactions.
Created Ping contract from Alice to Bob.
Transaction read: 0
Transaction read: 1
```

Now, this is due to the fact that I had already ran the code with `createFirstPing()` only. Maybe we should add a note about this, or instructions to reset the sandbox?

Note that you do do this later:
> Before running the application, start with a clean ledger to avoid receiving unprocessed contracts from previous examples:

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/ex-tutorial-nodejs/pull/7/commits/4fc8066183130afd874c1ff334d11e02e98346bf

stefanobaghino-da committed:
Address https://github.com/digital-asset/ex-tutorial-nodejs/pull/7#discussion_r287694730

stefanobaghino-da committed:
Address https://github.com/digital-asset/ex-tutorial-nodejs/pull/7#discussion_r287696301
